src/engraving/dom/engravingitem.cpp

```diff
@@ -1638,6 +1638,85 @@ RectF EngravingItem::symBbox(const SymIdList& symbols) const
     return score()->engravingFont()->bbox(symbols, magS());
 }
 
+Shape EngravingItem::symShape(SymId id) const
```

Comment:
I think better move this method to `IEngravingFont`

Reply:
And when doing that, we might want to cache the Shape calculated by this method, just like we do for the bbox actually

Reply:
> I think better move this method to `IEngravingFont`

Done!



> And when doing that, we might want to cache the Shape calculated by this method, just like we do for the bbox actually

I'm not sure I know where or how we would cache that Shape. How would you do this?

Reply:
> I'm not sure I know where or how we would cache that Shape. How would you do this?

You would add a new `shapeWithCutouts` field to the `EngravingFont::Sym` struct. This will probably need to be `mutable`.
Then, in `EngravingFont::shapeWithCutouts`, use `const Sym& s = sym(id)` to obtain the `Sym` for that symbol. Check if the `shapeWithCutouts` is empty; if not, just return it; if yes, then calculate it and store it for the next time, and return it.
You might also need to do some stuff with `useFallbackFont(id)`.

To some extent, `EngravingFont::bbox` is a good example, except that all bboxes are calculated in advance when the font is loaded, instead of just at the moment that they are needed.